### DISCUSSION TITLE
Using EOS module in Pressure-Temperature mode

### DISCUSSION THREAD

**Discussion by Quacken8** [MESAHub/mesa#514] [Q&A] [answered]
Hello,

I'm writing a thesis on solar structure based on one particular article ([Schüsser & Rempel 2005](http://arxiv.org/abs/astro-ph/0506654)) and I'd like to reproduce their results. It would be very useful for me to use MESA EOS in PT mode as that simplifies many of my caluclations significantly. 
In the sample_eos.f90 there's this one part that seems to suggest that MESA is able to run in PT mode:
```
! call the density-temperature based eos

! composition derivatives by isotope

         call eosDT_get( &
               handle, species, chem_id, net_iso, xa, &
               Rho, log10(Rho), T, log10(T), &
               res, d_dlnd, d_dlnT, d_dxa, ierr)


! report the results

         call PrettyOut(Rho, T, abar, zbar, res, d_dlnd, d_dlnT)


! example calling the pressure-temperature based eos
!         Pgas = exp(res(i_lnPgas))
!         call eosPT_get( &
!               handle, Z, X, abar, zbar, &
!               species, chem_id, net_iso, xa, &
!               Pgas, log10(Pgas), T, log10(T), &
!               Rho, log10Rho, dlnRho_dlnPgas_const_T, dlnRho_dlnT_const_Pgas, &
!               res, d_dlnd, d_dlnT, d_dabar, d_dzbar, ierr)
```
however it seems like this way was abandoned and I'm not savvy enough in fortran or thermodynamics yet to revive it. Can you help me make this module run in PT mode?

**Comment by rjfarmer** [comment, marked as answer]
eosPT_get still exists in the code https://github.com/MESAHub/mesa/blob/bb123bbb84441b37e399f12280a456c93b5dca19/eos/public/eos_lib.f90#L259 you just need to uncomment the code in the sample_eos.f90 file (and check that the order of the arguments still match)

**Reply by Quacken8** [reply]
Oh, cool! So the results it gives are consistent in terms of physics?

**Reply by rjfarmer** [reply]
Yes, under the hood eosPT is just eosDT but we handle the transformation between inputs/outputs for you. https://github.com/MESAHub/mesa/blob/80d1e8577e5bce0fff534e383795e578f61fc4c9/eos/private/eospt_eval.f90#L154

**Reply by Quacken8** [reply]
Amazing, thank you!

**Comment by Quacken8** [comment]
Follow up question: eos DT returns derivatives alongsite the resutls, eos_def names them d_dlnd and d_dlnT. When I run eos in PT mode, do these same arrays now hold (∂/∂lnP)_T and (∂/∂lnT)_P respectively?

**Comment by evbauer** [comment]
Not quite, but the routines should return enough information for you to construct what you need in terms of derivatives. In the code block where the public `eosPT_get` routine is declared, we can look through the variable declarations for the arguments to that routine to find the output variables:
https://github.com/MESAHub/mesa/blob/bb123bbb84441b37e399f12280a456c93b5dca19/eos/public/eos_lib.f90#L292-L307
This shows that you still have the usual vectors of derivatives such as `d_dlnRho_const_T`, and you also have a couple of additional derivative quantities `dlnRho_dlnPgas_const_T` and `dlnRho_dlnT_const_Pgas`, which you can use to construct (∂/∂lnP)_T and (∂/∂lnT)_P via the chain rule.
